staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,765 @@
+unsafe impl ByteValued for VirtioCanFrame {}
+
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+struct VirtioCanCtrlRequest {
+    msg_type: Le16,
+}
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioCanCtrlRequest {}
+
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+struct VirtioCanCtrlResponse {
+    result: i8,
+}
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioCanCtrlResponse {}
+
+pub(crate) struct VhostUserCanBackend {
+    controller: Arc<RwLock<CanController>>,
+    acked_features: u64,
+    event_idx: bool,
+    pub(crate) exit_event: EventFd,
+    mem: Option<GuestMemoryAtomic<GuestMemoryMmap>>,
+}
+
+type CanDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
+
+impl VhostUserCanBackend {
+    pub(crate) fn new(controller: Arc<RwLock<CanController>>) -> Result<Self> {
+        Ok(VhostUserCanBackend {
+            controller,
+            event_idx: false,
+            acked_features: 0x0,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+        })
+    }
+
+    fn check_features(&self, features: u16) -> bool {
+        (self.acked_features & (1 << features)) != 0
+    }
+
+    fn process_ctrl_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_ctrl_requests");
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if descriptors.is_empty() {
+                warn!("Error::UnexpectedDescriptorCount");
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            trace!("descriptors.len(): {:?}", descriptors.len());
+
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                warn!("Error::UnexpectedWriteOnlyDescriptor");
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            if desc_request.len() as usize != size_of::<VirtioCanCtrlRequest>() {
+                trace!("UnexpectedDescriptorSize, len = {:?}", desc_request.len());
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioCanCtrlRequest>(),
+                    desc_request.len(),
+                ));
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioCanCtrlRequest>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            match request.msg_type.into() {
+                VIRTIO_CAN_SET_CTRL_MODE_START => {
+                    //TODO: vcan->busoff = false;
```

Comment:
Is this not meant to update the status in the configuration register? The spec is not very clear about this.

Reply:
Also, I think the START/STOP should be implemented with a sort of state-machine so if the STOP msg is received in the STOPPED state, the response would be INVALID. The same would apply if the START is received in the STARTED state. I agree that the spec is not clear about this.

Reply:
I forgot to add that the spec states:
> The device MUST reject any CAN frame when CAN controller is in an invalid state with VIRTIO_CAN_RESULT_NOT_OK

Reply:
Thanks Matias, I am working on it, I will come back with a new version next days, aligned to spec's requirements.

Reply:
I think I have addressed all the above questions/suggestions:
>Is this not meant to update the status in the configuration register too? The spec is not very clear about this though.

The current implementation updates the status (config) of the device.

> Also, I think the START/STOP should be implemented with a sort of state-machine so if the STOP msg is received in the STOPPED state, the response would be INVALID. The same would apply if the START is received in the STARTED state. I agree that the spec is not clear about this. I think at least the status configuration register shall reflect that change.

 If the device is already started/stopped and receive a `VIRTIO_CAN_SET_CTRL_MODE_START`/`VIRTIO_CAN_SET_CTRL_MODE_STOP` ctrl message will reply with `VIRTIO_CAN_RESULT_NOT_OK`. Otherwise, will atempt to change the status (config) and return `VIRTIO_CAN_RESULT_OK`.

>The device MUST reject any CAN frame when CAN controller is in an invalid state with VIRTIO_CAN_RESULT_NOT_OK

When the controller is stopped or in invalid state rejects all CAN/FD frames (transmission, receiption)